src/library/scala/concurrent/Future.scala

```diff
@@ -413,8 +413,34 @@ trait Future[+T] extends Awaitable[T] {
    * @return        a `Future` with the result of the application of `f` to the results of `this` and `that`
    * @group Transformations
    */
-  def zipWith[U, R](that: Future[U])(f: (T, U) => R)(implicit executor: ExecutionContext): Future[R] =
-    flatMap(r1 => that.map(r2 => f(r1, r2)))(if (executor.isInstanceOf[BatchingExecutor]) executor else parasitic)
+  def zipWith[U, R](that: Future[U])(f: (T, U) => R)(implicit executor: ExecutionContext): Future[R] = {
+    val buffer = new AtomicReference[Success[Any]]()
+    val outPromise = Promise[R]()
+
+    this.onComplete{
+      case s: scala.util.Success[T] =>
+        val prevValue = buffer.getAndSet(s)
+        if (prevValue ne null) {
+          val oldValue = prevValue.get.asInstanceOf[U]
+          outPromise.complete(try Success(f(s.get, oldValue)) catch{case NonFatal(e) => Failure(e)})
```

Comment:
Using the extractor for NonFatal is not optimal for performance, so use the guard: `catch { case e if NonFatal(e) => Failure(e) }`

Reply:
got it, I changed it to a guard